operators/nftopology/controllers/packagerevision_controller.go

```diff
@@ -0,0 +1,258 @@
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Step 1: check package revision state
+	// TODO(s3wong): if package deleted, inform ESA
+	if pkgRev.Spec.Lifecycle != porchv1alpha1.PackageRevisionLifecyclePublished {
+		l.Info("Package lifecycle is not Published, ignore", "PackageRevision.lifecycle",
+			pkgRev.Spec.Lifecycle)
+		return ctrl.Result{}, nil
+	}
+
+	var regionName, clusterName, nfInstanceName, topologyName, topoNamespace string
+	var createDeployed, found bool
+	pkgRevLabels := pkgRev.ObjectMeta.Labels
+	if regionName, found = pkgRevLabels["region"]; !found {
+		l.Info("region name not found from package")
+		return ctrl.Result{}, nil
+	}
+	if nfInstanceName, found = pkgRevLabels["nfinstance"]; !found {
+		l.Info("NF Instance name not found from package")
+		return ctrl.Result{}, nil
+	}
+	if topologyName, found = pkgRevLabels["nftopology"]; !found {
+		l.Info("NF Topology name not found from package")
+		return ctrl.Result{}, nil
+	}
+	if topoNamespace, found = pkgRevLabels["nftopology-namespace"]; !found {
+		l.Info("NF Topology namespace not found from package")
+		return ctrl.Result{}, nil
+	}
+	// repo name is the cluster name
+	clusterName = pkgRev.Spec.RepositoryName
+
+	// Step 2: search for NFTopology and NFDeployed
+	nfTopology := &nfreqv1alpha1.NFTopology{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: topoNamespace, Name: topologyName}, nfTopology); err != nil {
+		l.Error(err, fmt.Sprintf("NFTopology object not found: %s\n", topologyName))
+		return ctrl.Result{}, nil
+	}
+
+	// TODO(s3wong): potentially, there would be multiple NFDeployedInstance CRs for an instance of
+	// NFInstance --- which means deploymentName should be unique
+	deploymentName := ConstructNFDeployedName(nfInstanceName, regionName, clusterName)
+
+	nfDeployed := &nfdeployv1alpha1.NFDeployed{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: topoNamespace, Name: nfTopology.ObjectMeta.Name}, nfDeployed); err != nil {
+		createDeployed = true
+		nfDeployed.ObjectMeta.Name = nfTopology.ObjectMeta.Name
+		nfDeployed.ObjectMeta.Namespace = nfTopology.ObjectMeta.Namespace
+	}
+
+	// Search for corresponding NFInstance
+	var nfInstance *nfreqv1alpha1.NFInstance = nil
+	for _, nfInst := range nfTopology.Spec.NFInstances {
+		if nfInst.Name == nfInstanceName {
+			nfInstance = &nfInst
+			break
+		}
+	}
+	if nfInstance == nil {
+		err := errors.New(fmt.Sprintf("%s not found in NF Topology %s\n", nfInstanceName, topologyName))
+		l.Error(err, fmt.Sprintf("NF Instance %s not found in NF Topology %s\n", nfInstanceName, topologyName))
+		return ctrl.Result{}, err
+	}
+	// Get NFClass, cluster scope
+	className := nfInstance.NFTemplate.ClassName
+	var nfClass = &nfreqv1alpha1.NFClass{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Name: className}, nfClass); err != nil {
+		l.Error(err, fmt.Sprintf("NFClass object not found: %s: %s\n", className, err.Error()))
+		return ctrl.Result{}, err
```

Comment:
I think this is fine for now, but eventually we may want to check things like this in a Validate method. What we do in a few other instances is we check references and stuff in Validate, and set the Stalled condition to True if there's dangling references.